service/src/main/java/bio/terra/workspace/service/iam/SamService.java

```diff
@@ -416,7 +448,55 @@ public void removeWorkspaceRole(
   }
 
   /**
-   * Wrapper around Sam client to retrieve the current permissions model of a workspace.
+   * Wrapper around the Sam client to remove a role from the provided user on a controlled resource.
+   *
+   * <p>Similar to {@removeWorkspaceRole}, but for controlled resources. This should only be
+   * necessary for private resources, as users do not have individual roles on shared resources.
+   *
+   * <p>This call to Sam is made as the WSM SA, as users do not have permission to directly modify
+   * IAM on resources. This method still requires user credentials to validate as a safeguard, but
+   * they are not used in the role removal call.
+   */
+  @Traced
+  public void removeResourceRole(
+      ControlledResource resource,
+      AuthenticatedUserRequest userRequest,
+      ControlledResourceIamRole role,
+      String email)
+      throws InterruptedException {
+    // Validate that the provided user credentials are an owner in the resource's workspace.
+    // Although the Sam call to revoke a resource role must use WSM SA credentials instead, this
+    // is a safeguard against accidentally invoking these credentials for unauthorized users.
+    checkAuthz(
+        userRequest,
+        SamConstants.SAM_WORKSPACE_RESOURCE,
+        resource.getWorkspaceId().toString(),
+        SamConstants.SAM_WORKSPACE_OWN_ACTION);
+
+    try {
+      ResourcesApi wsmSaResourceApi = samResourcesApi(getWsmServiceAccountToken());
+      SamRetry.retry(
+          () ->
+              wsmSaResourceApi.removeUserFromPolicyV2(
+                  resource.getCategory().getSamResourceName(),
+                  resource.getResourceId().toString(),
+                  role.toSamRole(),
+                  email));
+      logger.info(
+          "Removed role {} from user {} on resource {}",
+          role.toSamRole(),
+          email,
+          resource.getResourceId());
+    } catch (IOException credentialException) {
+      throw new InternalServerErrorException(
+          "Error removing resource role in Sam", credentialException);
```

Comment:
Seems like these two cases should have different error messages. Maybe say "internal error", though I guess the 500 gives it away.

Reply:
I think the error code should be the important bit here, but modified the error message to make it a little clearer as well.